target/riscv/csr.c

```diff
@@ -1237,19 +1233,34 @@ static int read_ccsr(CPURISCVState *env, int csrno, target_ulong *val)
     // The capability cause has moved to xTVAL so we don't report it here.
     RISCVCPU *cpu = env_archcpu(env);
     target_ulong ccsr = 0;
-    ccsr = set_field(ccsr, CCSR_ENABLE, cpu->cfg.ext_cheri);
-    ccsr = set_field(ccsr, CCSR_DIRTY, 1); /* Always report dirty */
+    ccsr = set_field(ccsr, XCCSR_ENABLE, cpu->cfg.ext_cheri);
+    ccsr = set_field(ccsr, XCCSR_DIRTY, 1); /* Always report dirty */
+
+    /* Expose state bits (GCLG) only above U mode */
+    if (env->priv != PRV_U)
+        ccsr |= env->mccsr;
+
     qemu_log_mask(CPU_LOG_INT, "Reading xCCSR(%#x): %x\n", csrno, (int)ccsr);
     *val = ccsr;
     return 0;
 }
 
 static int write_ccsr(CPURISCVState *env, int csrno, target_ulong val)
 {
-    error_report("Attempting to write " TARGET_FMT_lx
-                 "to xCCSR(%#x), this is not supported (yet?).",
-                 val, csrno);
-    return -1;
+    static const target_ulong gclgmask = (SCCSR_GCLGS | SCCSR_GCLGU);
+
+    /* U mode writes refused */
+    if (env->priv == PRV_U) {
+        error_report("Attempting to write " TARGET_FMT_lx
+                     "to xCCSR(%#x), this is not supported (yet?).",
+                     val, csrno);
+        return -1;
+    }
+
+    /* Take the GCLG bits from the store and update state bits */
+    env->mccsr = set_field(env->mccsr, gclgmask, get_field(val, gclgmask));
```

Comment:
Do we need another bit somewhere determining whether S-mode is allowed to write?

Reply:
No, they're like [ms]status.SIE. But they should really have S prefixes on all of them to make that clear.